services/streammanager/googlecloudfunction/googlecloudfunction.go

```diff
@@ -0,0 +1,215 @@
+	FunctionEnvironment    string `json:"functionEnvironment"`
+	RequireAuthentication  bool   `json:"requireAuthentication"`
+	GoogleCloudFunctionUrl string `json:"googleCloudFunctionUrl"`
+}
+
+type Client struct {
+	service *cloudfunctions.Service
+	opts    common.Opts
+}
+
+var pkgLogger logger.Logger
+
+func init() {
+	pkgLogger = logger.NewLogger().Child("streammanager").Child("GoogleCloudFunction")
+}
+
+type GoogleCloudFunctionProducer struct {
+	client *Client
+}
+
+// NewProducer creates a producer based on destination config
+func NewProducer(destination *backendconfig.DestinationT, o common.Opts) (*GoogleCloudFunctionProducer, error) {
+	var config Config
+	jsonConfig, err := json.Marshal(destination.Config)
+	if err != nil {
+		return nil, fmt.Errorf("[GoogleCloudFunction] Error while marshalling destination config :: %w", err)
+	}
+	err = json.Unmarshal(jsonConfig, &config)
+	if err != nil {
+		return nil, fmt.Errorf("[GoogleCloudFunction] error  :: error in GoogleCloudFunction while unmarshalling destination config:: %w", err)
+	}
+
+	opts := []option.ClientOption{
+		option.WithCredentialsJSON([]byte(config.Credentials)),
+	}
+
+	service, err := generateService(destination, opts...)
+	// If err is not nil then retrun
+	if err != nil {
+		pkgLogger.Errorf("[GoogleCloudFunction] error  :: %w", err)
+		return nil, err
+	}
+	client := &Client{service, o}
+
+	return &GoogleCloudFunctionProducer{client}, err
+}
+
+func (producer *GoogleCloudFunctionProducer) Produce(jsonData json.RawMessage, destConfig Config) (statusCode int, respStatus, responseMessage string) {
+
+	parsedJSON := gjson.ParseBytes(jsonData)
+	if destConfig.FunctionEnvironment == "gen1" && destConfig.RequireAuthentication {
+		client := producer.client
+
+		if client == nil {
+			respStatus = "Failure"
+			responseMessage = "[GoogleCloudFunction] error  :: Failed to initialize GoogleCloudFunction client"
+			return 400, respStatus, responseMessage
+		}
+
+		functionName := getFunctionName(destConfig.GoogleCloudFunctionUrl)
+
+		fmt.Print(functionName)
+
+		requestPayload := &cloudfunctions.CallFunctionRequest{
+			Data: `{"input_key": "input_value"}`,
+		}
+
+		// Make the HTTP request
+		call := client.service.Projects.Locations.Functions.Call(functionName, requestPayload)
+
+		response, err := call.Do()
+		if err != nil {
+			if apiErr, ok := err.(*googleapi.Error); ok && apiErr.Code == http.StatusNotModified {
+				fmt.Println("Function call was not executed (Not Modified)")
+			} else {
+				log.Fatalf("Failed to call function: %v", err)
+			}
+		}
+
+		// Process the response (sample response handling here).
+		if response != nil {
+			fmt.Printf("Function call status code: %d\n", response.HTTPStatusCode)
+			// Handle response content as needed.
+		}
+		fmt.Println("Request successful!")
+
+		respStatus = "Success"
+		responseMessage = "[GoogleCloudFunction] :: Message Payload inserted with messageId :: "
+		return 200, respStatus, responseMessage
```

Comment:
Can we put this in a separate function ?

Reply:
move gen1 and gen2 logic create 2 function @ujjwal-ab

Reply:
Done